Improve FileName editor widget

medspx opened the pull request:
This PR add some improvements to the File Name editor widget (the widget you use to easily add a file name in an attribute).

The main improvement is the ability to display an hyperlink on the file path. This hyperlink is clickable (with the OpenExternalLinks mechanism of Qt) and opens the file with the default viewer/software that is defined at the operating system level. Users will be able to open files directly from QGIS. Here is what it looks like:
![File Name widget with an hyperlink](https://cloud.githubusercontent.com/assets/8571115/11627072/2791b9da-9ce8-11e5-9038-eea0218aea0a.png)

I've also added some options to the widget (based on some use cases) and I've made a configuration dialog to set them :
![File Name edit widget configuration dialog box](https://cloud.githubusercontent.com/assets/8571115/11625336/91b7538c-9cda-11e5-9fd3-ec167969addc.png)

Here is a description of the options (there is a tooltip for each option):
- _Use an hyperlink for file name_: when checked the widget displays an hyperlink (in a QLabel) instead of a QLineEdit. The link is always clickable, even in non edit mode. In this mode, user can only change the content of the field by selecting a file with the file selector (which is opened with the "..." button). If you want to revert to editable QLineEdit file name widget, just unchek the option.
- _Display the full path_: when checked, the hyperlink will display the absolute path of the file. When unchecked, the hyperlink will display only the name of the file (useful if the file path is always long).
- _Default Path_: this is the path that will be used by the file selector when opening. The file selector will start at this path (if the option is not empty).
- _Store Relative paths to the default path_: when checked, file names are stored relatively to the default path described above. It is useful to store long paths into text shapefiles attributes limited to 254 characters.
- _Store relative paths to the project path_: when checked, file names are stored relatively to the project (.qgs file) path. This option will override the "save paths" setting of the project because it is sometimes useful to store code/svg/layer path relatively but not for the file name attributes.
- _File storage options_ are exclusive options to make the file selector select files (default mode) or only directories.

For options storage, I've made them non mandatory. They are stored in the project files only when they are set. The goal is to try to minimise the size of the project file (editor widgets configuration is stored twice in the .qgs file).

I've also made some fixes to the File Name editor widget:
- add a isFieldSupported method to the factory to allow File Name widget only on fields that are QStrings.
- add a new QSetting to save the last path used in the widget. Before the PR, users were always defaulted to QGIS home directory when using the file selector button. This QSetting could also be used for the photo widget.
- when you want to change the value of an already set attribute, file selector will point to this path.
- when the value of the attribute is null or non valid (not a file), the file selector will use the default path (if set), or the last used path or the home directory.

On the attribute table side, I've made nothing: the value displayed in the attribute table is the data stored in the attribute (so no hyperlink in the attribute table).

Commit:
Improve FileName editor widget

m-kuhn commented:
Great!

 * Would it make sense to group everything related to "default path" in a checkable group box? The reason is, "default path" will always be hard-coded (I guess?) and it should be possible to hide everything that's not relative away.
 * Use "a hyperlink" instead of "an hyperlink"
 * "Store only files" -> "Store file paths" (same for directories)
 * Semi related: widget configuration should really not be stored twice. Maybe that's related to a recent (post-2.12) change of mine.

nyalldawson commented:
@medspx good stuff!

medspx commented:
@m-kuhn 
For grouping, do you mean this:
![new dialog config](https://cloud.githubusercontent.com/assets/8571115/11652208/bc8d9da2-9d96-11e5-9550-f22a5e4faa9a.png)
with "Store paths relatives to the project path" set by default and "Store paths relatives to the default path" enabled only if default path is non empty .?

For "hard coded" default path, do you mean I could use an expression/variable ?

For the twice stored widget configuration, I have seen this since at least 2.4. There are two <edittypes> blocks in the .qgs file. The first is just after the map-layer-style-manager at the beginning of the maplayer definition. The second is at the end of the maplayer definition (after <conditionalstyles> tag). When you have tables with lots of attributes the size of the project file willis quite big (I have a .qgs file with a size > 3 Mb). I haven't found anything in the bugtracker related to this. Should I open a ticket ?

m-kuhn commented:
Is the default path always required? If not, I would move the widgets related ot the default path below the "Store parths relative* to the detault path" and only enable them when they are required.

* btw, there's a typo with plural

m-kuhn commented:
And yes, please open an issue with a small example.

medspx commented:
Default path is a way to make the file selector of the edit widget to point to the same base directory for each feature. It is not mandatory: you can set it or not (if you want the file selector in the edit widget to point only on last used path).

It is quite independent from "relative paths" ("Store relative paths to default path" use it only if it is set): you can have a default path with an absolute paths storage and you can have no default path and a relative paths storage (on the project). That's why I've not grouped them at first. The only "link" between default path and relative paths is the option to use the default path as the base root of the edit widget file selector by toggling "Store relative paths to default path".

If I put "default path" widgets below the "store paths relative to default path" it means that "default path" can only be set if "store paths relative to default path" is toggled which is problematic as I said before that user would like to have a base default path for the edit widget file selector without having a relative storage option. Is it clearer now and what can you advice to me ?

medspx commented:
The [bug report 13960](https://hub.qgis.org/issues/13960) is opened !

m-kuhn commented:
It's maybe a UX thing. If I see an empty field, I normally want to fill something in and feel uncomfortable if I don't.
Maybe you could add an explanatory placeholder text which will appear if it is empty? If that explains what happens if the field is left empty, that should be fine.

Commit:
Improve FileName editor widget (fix config dialog)

m-kuhn commented:
"File selection will start from this place" is not exactly what happens if left empty, no? ;)

m-kuhn commented:
@medspx What do you think about this:

By default the project path will be used as "default path" (and written as placeholder).
Once the user used the file widget and navigated to another path, the other path will be used as default path (and written as placeholder).
The user can insert a "default path" which will overwrite this default behavior and be used as "fixed default path". Only when the user enters something the option "Store paths relative to default path" will be available.

Commit:
Improve FileName widget editor configuration

medspx commented:
@m-kuhn 
You are right, that's a very good solution. I have made the changes in the last commit. I have also updated the tooltips to inform user about the default path config option and fixed the widget editor file selection.

Commit:
Fix FileName widget editor configuration

3nids commented:
I am was working on refactoring a bit the file name widget, while I discovered this PR.

My ideas are:

* create a proper class QgsFileNameWidget (to have the code in its dedicated class, and to be able to have a custom widget plugin for it*)
* merge the photo widget functionality for it (display an image instead of a link/file path)
* configuration of the widget determines what is shown (line edit or link or photo)

Hence both QgsFileNameWidgetWrapper and QgsPhotoWidgetWrapper would use the same class for the widget.

@medspx please let me know what you think. Would you like to do these changes? Otherwise, I can do it. If I do it, @m-kuhn let me know if it's ok to merge, and I'll put my changes on top!

* look at QgsRelationReferenceWidget for instance

m-kuhn commented:
Good from my side.

I have some ideas for the file name widget ;) like connecting it to a [`QFileSystemModel`](http://doc.qt.io/qt-4.8/qfilesystemmodel.html) but make sure to keep in mind that http or other schemas need to be supported as well.

3nids commented:
@m-kuhn can I merge this, then?
I will do the changes later on, on top

medspx commented:
@3nids Well, I think your changes and your idea of going further with the File name widget are welcomed ! Combining File Name and Photo widgets is also very clever. I thought to work on Photo widget after this PR would be merged, mainly to make it use the new lastFileNamePath QSetting. Photo widget needs also to support better default auto-size.

For the moment, I am working on something else so there is no problem for you to add your code.

m-kuhn commented:
Please make it pass on travis before merging. It's only a formatting issue @medspx see here https://github.com/m-kuhn/QGIS-Website/blob/qtcreator.rst/source/site/getinvolved/development/git.rst#setup-pre-commit-hook

@3nids can you also cross check if it works as expected?

3nids commented:
ok, I will take care of both.
thanks

3nids commented on src/gui/editorwidgets/qgsfilenamewidgetwrapper.cpp:
```diff
@@ -101,42 +136,132 @@ void QgsFileNameWidgetWrapper::initWidget( QWidget* editor )
+  else
+  {
+    // If there is no valid value, find a default path to use
+    QUrl theUrl = QUrl::fromUserInput( oldPath );
```
oldPath is never assigned here, is it expected?

medspx replied:
No, it is a mistake from code manipulations. oldPath is assigned NULL (well, [I think](https://stackoverflow.com/questions/9724966/how-to-initialize-a-qstring-to-null)).

The final else statement should be deleted and its content put outside of the initial test: we always need to evaluate if oldpath is valid or not (and not evaluate it only when the value is null). Would you like to fix this (or should I do it) ?

3nids replied:
Thanks for the feedback.
I will try to fix, I'll show it to you asap.

3nids commented:
superseeds #2589
